driver/src/main/com/mongodb/GridFSBucketImpl.java

```diff
@@ -0,0 +1,316 @@
+    }
+
+    @Override
+    public void downloadToStreamByName(final String filename, final OutputStream destination) {
+        downloadToStreamByName(filename, destination, new GridFSDownloadByNameOptions());
+    }
+
+    @Override
+    public void downloadToStreamByName(final String filename, final OutputStream destination, final GridFSDownloadByNameOptions options) {
+        downloadToStream(openDownloadStreamByName(filename, options), destination);
+    }
+
+    @Override
+    public GridFSFindIterable<Document> find() {
+        return new GridFSFindIterableImpl<Document>(getFilesCollection().find());
+    }
+
+    @Override
+    public <TResult> GridFSFindIterable<TResult> find(final Class<TResult> resultClass) {
+        return new GridFSFindIterableImpl<TResult>(getFilesCollection(resultClass).find());
+    }
+
+    @Override
+    public GridFSFindIterable<Document> find(final Bson filter) {
+        return find().filter(filter);
+    }
+
+    @Override
+    public <TResult> GridFSFindIterable<TResult> find(final Bson filter, final Class<TResult> resultClass) {
+        return find(resultClass).filter(filter);
+    }
+
+    @Override
+    public void delete(final ObjectId id) {
+        DeleteResult result = filesCollection.deleteOne(new BsonDocument("_id", new BsonObjectId(id)));
+        chunksCollection.deleteMany(new BsonDocument("files_id", new BsonObjectId(id)));
+
+        if (result.wasAcknowledged() && result.getDeletedCount() == 0) {
+            throw new MongoGridFSException(format("No file found with the ObjectId: %s", id));
+        }
+    }
+
+    @Override
+    public void rename(final ObjectId id, final String newFilename) {
+        UpdateResult updateResult = filesCollection.updateOne(new BsonDocument("_id", new BsonObjectId(id)),
+                new BsonDocument("$set", new BsonDocument("filename", new BsonString(newFilename))));
+
+        if (updateResult.wasAcknowledged() && updateResult.getMatchedCount() == 0) {
+            throw new MongoGridFSException(format("No file found with the ObjectId: %s", id));
+        }
+    }
+
+    private MongoCollection<Document> getFilesCollection() {
+        return getFilesCollection(Document.class);
+    }
+
+    private <TResult> MongoCollection<TResult> getFilesCollection(final Class<TResult> clazz) {
+        return database.getCollection(bucketName + ".files", clazz)
+                .withCodecRegistry(codecRegistry)
+                .withReadPreference(readPreference)
+                .withWriteConcern(writeConcern);
+    }
+
+    private <TResult> MongoCollection<TResult> getChunksCollection(final Class<TResult> clazz) {
+        return database.getCollection(bucketName + ".chunks", clazz)
+                .withCodecRegistry(codecRegistry)
+                .withReadPreference(readPreference)
+                .withWriteConcern(writeConcern);
+    }
+
+    private void checkCreateIndex() {
+        if (!checkedIndexes) {
```

Comment:
checkedIndex needs to be volatile to ensure that the value is published.  See JCIP 3.1.4